babel/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -147,67 +147,67 @@ boolean IsNullable() :
     )
 }
 
-boolean IsCaseSpecific() :
+SqlColumnAttribute ColumnAttributeCaseSpecific() :
 {
-
+    boolean isCaseSpecific = true;
 }
 {
-    (
-        <NOT> <CASESPECIFIC> {
-            return false;
-        }
-    |
-        <CASESPECIFIC> {
-            return true;
-        }
-    )
+    [ <NOT> { isCaseSpecific = false; } ]
+    <CASESPECIFIC> {
+        return new SqlColumnAttributeCaseSpecific(getPos(), isCaseSpecific);
+    }
 }
 
-boolean IsUpperCase() :
+SqlColumnAttribute ColumnAttributeUpperCase() :
+{
+    boolean isUpperCase = true;
+}
 {
+    [ <NOT> { isUpperCase = false; } ]
+    <UPPERCASE> {
+        return new SqlColumnAttributeUpperCase(getPos(), isUpperCase);
+    }
+}
 
+void ColumnAttributes(List<SqlColumnAttribute> list) :
+{
+    SqlColumnAttribute e;
+    Span s;
 }
 {
     (
-        <NOT> <UPPERCASE> {
-            return false;
-        }
-    |
-        <UPPERCASE> {
-            return true;
-        }
-    )
+        (
+            e = ColumnAttributeUpperCase()
+        |
+            e = ColumnAttributeCaseSpecific()
+        ) { list.add(e); }
+    )+
 }
 
 void ColumnWithType(List<SqlNode> list) :
 {
     SqlIdentifier id;
     SqlDataTypeSpec type;
     boolean nullable = true;
-    Boolean uppercase = null;
-    Boolean caseSpecific = null;
+    List<SqlColumnAttribute> columnAttributes =
+        new ArrayList<SqlColumnAttribute>();
     final Span s = Span.of();
 }
 {
     id = CompoundIdentifier()
     type = DataType()
-    // This acts as a loop to check which optional parameters have been specified.
+    /* This structure is to support [NOT] NULL appearing anywhere in the
```

Comment:
How come nullable is not also a ColumnAttribute class? Would converting it to one make it easier to parse all the column attributes in a loop rather than having a seperate boolean for null?

Reply:
Nullable is a special case and is referenced in quite a few places. I spoke with Drew and we decided that it was best to just leave it how it was.